[CAM] Fix ramp dressup performance

davidgilkaufman opened the pull request:
Sub-project 3 of [grant proposal](https://github.com/FreeCAD/FPA-grant-proposals/issues/46)

TODO create issue

Sometimes, when ramping long paths, FreeCAD freezes during ramp path computation, and can remain frozen for minutes. The problem can be seen on this sample file: [sample_model.zip](https://github.com/user-attachments/files/20699787/sample_model.zip). The generated gcode for this model does not change with this PR, but computation speeds up (on my machine) from ~2 minutes to 1.5 seconds.

The two biggest problems were:

- a call to edges.index(edge) to find the index in an array. I replaced this with tracking the index during iteration
- looping over an array of rapid edges to check if the current edge is a rapid move. When this is done on input edges I replaced this with index tracking as well, and when it is done on output edges I replaced it with tagging output edges as rapid/normal at generation time

The other problem I fixed was time wasted in unused Path.Log.debug calls. Even if the log line is not kept, these lines take time to invoke traceback.extract_stack and determine the current file (and thus the log level of the current file), and if this is done in a sufficiently hot loop, that time matters. I commented out a few notably influential log calls, and some others that were less notably problematic as well.

After these changes, profiling ([python line profiler](https://github.com/pyutils/line_profiler)) shows that the remaining wasted time is largely spent doing small property lookups and method calls from python into C inside busy loops. This could be further improved by operating on a path representation that stays closer to gcode/python, but I decided that the current state is good enough and it's not worth rewriting that much code for further improvements at the moment.

Here is some sample profiler output after performance improvements. Note that the most expensive remaining lines are high level function calls (the real work is inside somewhere) and python property access:

```
$ cat profile_output.txt | sort -n -k 3
# ...(all entries remaining over 50ms:)
# line   call count    time(ns)   per call (ns)
   282      1578   58881149.0  37313.8      7.2                          cp1 = candidate.Vertexes[1].Point
   292      5989   63172715.0  10548.1     13.5      pt = edge.valueAt(edge.LastParameter) if not flip else edge.valueAt(edge.FirstParameter)
   246      1848   68927189.0  37298.3      8.4                  p1 = edge.Vertexes[1].Point
   294      5989   71505047.0  11939.4     15.3      if type(edge.Curve) == Part.Line or type(edge.Curve) == Part.LineSegment:
   499      2315   72107705.0  31148.0     38.8              edge = edgeForCmd(cmd, startPoint)
   245      1848   74985932.0  40576.8      9.1                  p0 = edge.Vertexes[0].Point
   281      1578   80796801.0  51202.0      9.8                          cp0 = candidate.Vertexes[0].Point
   508         1  101011844.0    1e+08     54.3      return (Part.Wire(edges), rapid, rapid_indexes)
   562       221  112466102.0 508896.4     27.0                  rampedges = self.getreversed(rampedges)
   556      3178  138311942.0  43521.7     33.1                      outedges.append(self.createRampEdge(redge, curPoint, newPoint))
   228         2  194172169.0    1e+08     10.1          self.wire, self.rapids, self.rapid_indexes = Path.Geom.wireForPath(
   316       154  432265112.0    3e+06     52.6                                  self.createRampMethod1(rampedges, p0, projectionlen, rampangle)
   767      5989  551057905.0  92011.7     76.5                  commands.extend(Path.Geom.cmdsForEdge(edge))
   232         1  833127784.0    8e+08     43.4              self.outedges = self.generateRamps()
   235         1  868352735.0    9e+08     45.2          obj.Path = self.createCommands(obj, self.outedges)
```

tarman3 commented:
RampEntry  Dressup add useless or/and in some cases dangerous movement to `X0Y0`
<img width="650" height="264" alt="Screenshot_20250906_080622_lossy" src="https://github.com/user-attachments/assets/603f2fa9-bab6-4cbe-b5a3-ecd2a4b7601a" />

It was PR #22484 to fix this 
but now logic of this dressup completely changed

Can you try to find new solution to fix this problem?

davidgilkaufman commented:
I'm sorry I created a terrible merge conflict for you.

In an early version of my code I actually accidentally eliminated the X0Y0 you're talking about, but I added more code to bring it back because I was aiming to make a performance change while preserving functionality as well as possible. Take a look at RampEntry.py lines 360-366 -- that's the code I added to preserve the existing X0Y0 behavior, and I think you can achieve your desired behavior by deleting those lines.

For what it's worth, I think removing X0Y0 in that line is an improvement, but I'm not sure if I'm missing any important use cases/problems. I suspect a PR to change this will not merge before the upcoming release, but please make the PR and I think we should accept it.
